src/schedule/zephyr_ll.c

```diff
@@ -0,0 +1,296 @@
+// SPDX-License-Identifier: BSD-3-Clause
+//
+// Copyright(c) 2021 Intel Corporation. All rights reserved.
+//
+// Author: Guennadi Liakhovetski <guennadi.liakhovetski@linux.intel.com>
+
+#include <sof/list.h>
+#include <sof/spinlock.h>
+#include <sof/audio/component.h>
+#include <sof/schedule/ll_schedule_domain.h>
+#include <sof/schedule/schedule.h>
+#include <sof/schedule/task.h>
+
+/* per-scheduler data */
+struct zephyr_ll {
+	struct list_item tasks;			/* list of ll tasks */
+	spinlock_t lock;			/* protects the list or tasks */
+	struct ll_schedule_domain *ll_domain;	/* scheduling domain */
+};
+
+/* per-task scheduler data: so far actually unused */
+struct zephyr_ll_pdata {
+	uint64_t period;
+};
+
+//FIXME
+#define domain_time(sch) platform_timer_get_atomic(timer_get())
+
+static void zephyr_ll_run(void *data)
+{
+	struct zephyr_ll *sch = data;
+	struct task *task;
+	struct list_item *list, *tmp;
+
+// FIXME: locking
+	/*
+	 * We cannot use a mutex because tasks can be added or deleted from
+	 * interrupt context. TODO: use two lists: staging and running. Use the
+	 * lock to access the staging list, only access the running list in this
+	 * function - with no locking. To cancel set the state to CANCEL, here
+	 * check before running with lock held. Then release the lock and run
+	 * the task. If the task is cancelled during running - that's normal.
+	 * After task return take the lock, if cancelled - remove from the list
+	 * and set to FREE.
+	 */
+	spin_lock(&sch->lock);
+	list_for_item_safe(list, tmp, &sch->tasks) {
+		struct comp_dev *sched_comp;
+		enum task_state state;
+		uint32_t flags;
+
+		task = container_of(list, struct task, list);
+
+		if (task->state == SOF_TASK_STATE_CANCEL) {
+			list_item_del(&task->list);
+			task->state = SOF_TASK_STATE_FREE;
+			continue;
+		}
+
+		if (!domain_is_pending(sch->ll_domain, task, &sched_comp))
+			continue;
+
+		task->state = SOF_TASK_STATE_RUNNING;
+		state = task_run(task);
```

Comment:
We should really release the spinlock before we run() and then acquire it again after the run_task() returns.

Reply:
we have to research well what contexts can race for which resources here. IIUC and scheduling can be triggered from interrupt handlers, then the list of tasks has to be protected. In that case we cannot just drop the lock in the middle of this loop. One solution would be what I proposed - first extract all runnable tasks from the global list into a local one, while holding the lock. Then run the tasks from the local list, with no locking this time. The disadvantage would be, that tasks, returning `RESCHEDULE` have to be re-added to the global list, but I don't have a better solution so far.